### ISSUE TITLE
Build fails but Build pod completes successfully

### ISSUE THREAD

**Issue by kalgary** [openshift/origin#21154]
Our jenkins pipeline fails from time to time to build the application. The build exits after 0 sec throwing an Error but the build pod completes successfully.

The build is configured to run on 3 of our working nodes, there are no specific information about the reason of the failure neither in the nodes logs nor in the build logs. 

##### Version

oc v3.9.0+191fece
kubernetes v1.9.1+a0ce1bc657
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://master.test.cloud.cp.local:8443
openshift v3.7.1+c2ce2c0-1
kubernetes v1.7.6+a08f5eeb62

##### Steps To Reproduce

1. Run the jenkins pipeline 

$ oc start-build bc/XXX-pipeline

##### Current Result

Pipeline fails with an Error state despite the build pod completed successfully.

$ oc get pods

XXX-34-build    0/1       Completed    0          34m

$ oc get builds 

XXX-34                  Docker            Git@5db6904   Error

##### Expected Result

Pipeline runs succesful.

##### Additional Information

$ oc logs build/XXX-34 --loglevel=10

I1003 04:31:30.341225    4746 request.go:874] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"build XXX-34 is in an error state. No logs are available.","reason":"BadRequest","code":400}
I1003 04:31:30.341450    4746 helpers.go:201] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "build XXX-34 is in an error state. No logs are available.",
  "reason": "BadRequest",
  "code": 400
}]
F1003 04:31:30.341466    4746 helpers.go:119] Error from server (BadRequest): build XXX-34 is in an error state. No logs are available.

**Jenkins Stages:**

stage('Maven') {
      steps {
        echo 'Start MVN build'
        sh 'mvn --update-snapshots clean verify'
        echo 'archiving artifacts to Jenkins master'
        archiveArtifacts artifacts: '**/target/*.war', fingerprint: true
      }
    }

    stage('Image') {
      steps {
        sh 'printenv | sort'
        echo 'trigger openshift docker build ...'
        // Jenkins 2.73:
        openshiftBuild bldCfg: "${env.APP_NAME}", namespace: "${env.PROJECT_NAME}", env: [[ name: "ARTIFACT_URL", value: "${JENKINS_URL}job/${JOB_NAME}/lastBuild/artifact/target/quotation.war" ]]

        sleep time: 10, unit: "SECONDS"

        echo 'wait for docker build to complete ... '
        openshiftVerifyBuild bldCfg: "${env.APP_NAME}", namespace: "${env.PROJECT_NAME}", waitTime: '20', waitUnit: 'min'
      }
    }

$ oc edit bc XXX

apiVersion: build.openshift.io/v1
kind: BuildConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"BuildConfig","metadata":{"annotations":{},"labels":{"app":"XXX"},"name":"XXX","namespace":"----XXX"},"spec":{"output":{"to":{"kind":"ImageStreamTag","name":"XXX:latest"}},"postCommit":{},"resources":{},"runPolicy":"Serial","source":{"contextDir":"src/main/setup/kubernetes/image","git":{"ref":"develop","uri":"https://git.dev.+++/r/XXX/XXX.git"},"type":"Docker"},"strategy":{"dockerStrategy":{"env":[{"name":"GIT_SSL_NO_VERIFY","value":"true"}]},"type":"Docker"},"triggers":[]}}
  creationTimestamp: 2018-08-01T09:23:13Z
  labels:
    app: XXX
  name: XXX
  namespace: ----XXX
  resourceVersion: "74108076"
  selfLink: /apis/build.openshift.io/v1/namespaces/----XXX/buildconfigs/XXX
  uid: 83bea1d0-956c-11e8-9f01-005056be20f3
spec:
  nodeSelector: null
  output:
    to:
      kind: ImageStreamTag
      name: XXX:latest
  postCommit: {}
  resources: {}
  runPolicy: Serial
  source:
    contextDir: src/main/setup/kubernetes/image
    git:
      ref: develop
      uri: https://git.dev.+++/r/XXX/XXX.git
    type: Git
  strategy:
    dockerStrategy:
      env:
      - name: GIT_SSL_NO_VERIFY
        value: "true"
      - name: BUILD_LOGLEVEL
        value: "10"
    type: Docker
  triggers: []
status:
  lastVersion: 37

**Comment by jwforres** [comment]
@openshift/sig-developer-experience

**Comment by gabemontero** [comment]
@kalgary - need some help with the details you provided

I could use 
- the complete yaml (where you redact as needed) the pipeline build object, docker build object, and docker pod.

- the jenkins console logs for the job that failed, and the jenkins pod log when the failure occurs

- on the oc logs --loglevel=10 error you noted, that looks more like a server/controller side situation ... is that the only output you got?  If not, could your provide the complete output?  

Also, @bparees - you think the build controller or master logs would be beneficial in sorting out 

```
I1003 04:31:30.341225 4746 request.go:874] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"build XXX-34 is in an error state. No logs are available.","reason":"BadRequest","code":400}
I1003 04:31:30.341450 4746 helpers.go:201] server response object: [{
"metadata": {},
"status": "Failure",
"message": "build XXX-34 is in an error state. No logs are available.",
"reason": "BadRequest",
"code": 400
}]
F1003 04:31:30.341466 4746 helpers.go:119] Error from server (BadRequest): build XXX-34 is in an error state. No logs are available.
```
from the `oc logs` invocation @kalgary noted?

**Comment by bparees** [comment]
>  Also, @bparees - you think the build controller or master logs would be beneficial in sorting out

First i think getting the build yaml and build pod yaml would be useful.

usually builds end up in an error state when the build pod is deleted or terminates abnormally.  It would be good to understand what happened to this one.  Especially since @kalgary claims the build pod completed successfully.

**Comment by kalgary** [comment]
Hi guys, 
thanks, here the files that you have requested:

[build-34.txt](https://github.com/openshift/origin/files/2477647/build-34.txt)
[build-obj.txt](https://github.com/openshift/origin/files/2477667/build-obj.txt)
[build-pipeline.txt](https://github.com/openshift/origin/files/2477668/build-pipeline.txt)
 
As already descreibed the pipeline is started by a build (build-pipeline), the pipeline has been provided above, afterwards the stage('Image') runs another build (build-obj.txt) which is the one that, according to openshift, has failed. As you can see in the logs (build-34.txt), the build has completed successfully.

I cannot provide you any Jenkins logs since they were deleted, and I cannot reproduce the issue easily, since it randomly occurs. I can tell you that the jenkins logs notify that the build object failed. 

The build object is also marked as failed 

build-34                  Docker            Git@5db6904   Error 

Thanks in advance for your help.

**Comment by bparees** [comment]
I'm afraid that doesn't look like what we requested. You appear to have given us:

1) a partial build log
2) the buildconfig yaml (not the build yaml and not the build pod yaml)

what you need to retrieve:

```
oc get your-build-pod-name -o yaml
oc get build/your-build-pod-name -o yaml
oc logs build/your-build-name
```

For the build that is being triggered by the pipeline.

**Comment by bparees** [comment]
build-34.txt is clearly not a complete build log file.  It starts off with:

```
 October 3rd 2018, 09:41:15.873  Pushed 16/17 layers, 94% complete
```

Nor does it end as a normal build log would end.

The build pod yaml you attached shows the build is in fact still pending, it did not complete successfully (or fail).

```
status:
  phase: Pending
  qosClass: BestEffort
```

And you have still not provided a build yaml.

build-34 - this is a truncated build log file.
build-obj - this is a buildconfig yaml, not a build yaml
build-pipeline - this is a buildconfig for your pipeline build

**Comment by kalgary** [comment]
@bparees 
please read the build log file bottom up. this is a copy paste from kibana (I cannot retrieve the build logs form console).

Those are anyway the missing lines:

```
Pushed 18/18 layers, 100% complete
Push successful
```

here the build yaml:
[build.txt](https://github.com/openshift/origin/files/2492401/build.txt)


thanks.

**Comment by bparees** [comment]
running your builds with --build-loglevel=5 may provide more insight, other than that there isn't much i can currently tell you.  The build clearly shows it is in `phase: Error` but there is not enough information to determine why the build is in that phase.

It would also be good to understand if you can reproduce this w/o using jenkins/pipelines.  e.g. if you simply run "oc start-build" for your build, does the build ever end in an Error state?

**Comment by kalgary** [comment]
I know, I have debugged the issue, but I did not find any hint to what's going on. 
It seems that the build pod is registered as failed but it is not. 

Is it possible that some network issue prevent the correct update of the build status ? 
It also interesting to notice that this problem did not occure since days now !

I'm just sorry that I cannot provide you more insights, since any time a developer ran into this problem he/she re ran the pipeline deleting the erroneus build.

Anyway --build-loglevel=5 is not helpful in this case :(. 

As soon as the issue will arise again, I will try to collect all the info whcih you have asked for.

thanks.

**Comment by bparees** [comment]
the other useful data would be to run your master/controller process with loglevel 5 and gather the controller logs from the time period covering the failure.

**Comment by kalgary** [comment]
Hi,
Just in case anyone run into the same problem.

As suggested by @bparees I run my masters api and controller process with loglevel 5 and here what I found: 

Nov 06 12:28:18 master03.XXX.local origin-master-controllers[10285]: I1106 12:28:18.483978   10285 build_controller.go:1100] Patching build release-XXX/YYY-8 (New) with buildUpdate(phase: "Error", reason: "BuildPodExists", message: "The pod for this build already exists and is older than the build.", startTime: "2018-11-06 12:28:18.483600402 +0100 CET m=+83893.315201897", completionTime: "2018-11-06 12:28:18.483600402 +0100 CET m=+83893.315201897", duration: "0s", podName: "YYY-8-build")

I checked the Start Time of the build : 

```
oc describe bc oms-frontend
1:08 PM
YYY-8  error   0s              2018-11-06 **06:41:34** -0500 EST 
```
 and if compared with the one from the build pod : 

```
oc describe pod YYY-8-build
1:09 PM
Started:      Tue, 06 Nov 2018 **06:41:23** -0500
```
It looked like the build was indeed created before the build pod.

The reason was a time difference between master03 and nodes ! The problem was solved after a time sync. 

Thanks for the help!
